src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,63 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+        if (varTypeIsStruct(compiler->info.compRetType))
+        {
+            if (compiler->compMethodReturnsMultiRegRetType())
+            {
+                // Get the return tye of the struct.
+                SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR retStructDesc;
+                compiler->eeGetSystemVAmd64PassStructInRegisterDescriptor(compiler->info.compMethodInfo->args.retTypeClass, &retStructDesc);
+
+                assert(retStructDesc.passedInRegisters);
+                // In case of a single register returned struct , the return type have been normalized.
+                // In such case the execution will not be here 
+                // (the if (varTypeIsStruct(compiler->info.compRetType)) will evaluate to false).
+                assert(retStructDesc.eightByteCount > 1 &&
+                    retStructDesc.eightByteCount <= CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
```

Comment:
@sivarv, or is your point that the code below is already assuming 2?  If so, then perhaps it makes sense to also assert that it equals 2 in case CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS ever changes...not that it ever would...